# Credit Bureau Accountability Scorecard

## 1. Setup & Imports

In [1]:
# Install driver
!pip3 install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


In [2]:
!pip3 install sqlalchemy psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


In [3]:
# Connect to PostgreSQL
from sqlalchemy import create_engine
import pandas as pd

## 2. Load Data from PostgreSQL

In [4]:
engine = create_engine('postgresql://postgres:password@localhost:5432/postgres')

In [5]:
df = pd.read_sql('SELECT * FROM complaints', engine)
print(df.shape)
df.head()

(1282355, 18)


,date_received,product,sub_product,issue,sub_issue,consumer_narrative,company_public_response,company,state,zip_code,tags,consumer_consent,submitted_via,date_sent,company_response,timely_response,consumer_disputed,complaint_id
0,05/10/2019,Checking or savings account,Checking account,Managing an account,Problem using a debit or ATM card,None,None,NAVY FEDERAL CREDIT UNION,FL,328XX,Older American,None,Web,05/10/2019,In progress,Yes,N/A,3238275
1,05/10/2019,Checking or savings account,Other banking product or service,Managing an account,Deposits and withdrawals,None,None,BOEING EMPLOYEES CREDIT UNION,WA,98204,None,N/A,Referral,05/10/2019,Closed with explanation,Yes,N/A,3238228
2,05/10/2019,Debt collection,Payday loan debt,Communication tactics,Frequent or repeated calls,None,None,CURO Intermediate Holdings,TX,751XX,None,None,Web,05/10/2019,Closed with explanation,Yes,N/A,3237964
3,05/10/2019,"Credit reporting, credit repair services, or o...",Credit reporting,Incorrect information on your report,Old information reappears or never goes away,None,None,Ad Astra Recovery Services Inc,LA,708XX,None,None,Web,05/10/2019,Closed with explanation,Yes,N/A,3238479
4,05/10/2019,Checking or savings account,Checking account,Managing an account,Banking errors,None,None,ALLY FINANCIAL INC.,AZ,85205,None,N/A,Postal mail,05/10/2019,In progress,Yes,N/A,3238460


## 3. Data Cleaning & Transformation

### 3.1 Rename Columns to Snake Case

In [6]:
df.columns.tolist()

['date_received',
 'product',
 'sub_product',
 'issue',
 'sub_issue',
 'consumer_narrative',
 'company_public_response',
 'company',
 'state',
 'zip_code',
 'tags',
 'consumer_consent',
 'submitted_via',
 'date_sent',
 'company_response',
 'timely_response',
 'consumer_disputed',
 'complaint_id']

In [7]:
# Rename columns on the full dataset
df.columns = [
    'date_received',
    'product',
    'sub_product',
    'issue',
    'sub_issue',
    'consumer_narrative',
    'company_public_response',
    'company',
    'state',
    'zip_code',
    'tags',
    'consumer_consent',
    'submitted_via',
    'date_sent',
    'company_response',
    'timely_response',
    'consumer_disputed',
    'complaint_id'
]

df.sample(10)

,date_received,product,sub_product,issue,sub_issue,consumer_narrative,company_public_response,company,state,zip_code,tags,consumer_consent,submitted_via,date_sent,company_response,timely_response,consumer_disputed,complaint_id
10691,04/17/2019,Debt collection,Other debt,Attempts to collect debt not owed,Debt is not yours,None,None,"Weber & Olcese, PLC",MI,495XX,None,None,Web,04/17/2019,Closed with explanation,Yes,N/A,3215099
265605,04/10/2018,"Credit reporting, credit repair services, or o...",Credit reporting,Improper use of your report,Credit inquiries on your report that you don't...,None,Company has responded to the consumer and the ...,Experian Information Solutions Inc.,FL,33426,None,Consent not provided,Web,04/10/2018,Closed with non-monetary relief,Yes,N/A,2870261
408877,09/21/2017,"Credit reporting, credit repair services, or o...",Credit reporting,Problem with fraud alerts or security freezes,None,None,Company has responded to the consumer and the ...,Experian Information Solutions Inc.,FL,33411,None,Consent not provided,Web,09/21/2017,Closed with explanation,Yes,N/A,2680368
231789,05/24/2018,Debt collection,Other debt,False statements or representation,Attempted to collect wrong amount,In 2012 I moved out of an apartment. Before I ...,None,"RentDebt Automated Collections, LLC",KY,None,None,Consent provided,Web,05/24/2018,Closed with explanation,Yes,N/A,2916993
923694,02/23/2015,Bank account or service,Other bank product/service,Problems caused by my funds being low,None,None,None,FIFTH THIRD FINANCIAL CORPORATION,KY,40218,None,N/A,Referral,02/25/2015,Closed with explanation,Yes,No,1251759
889410,05/07/2015,Credit reporting,None,Incorrect information on credit report,Account status,None,None,CAPITAL ONE FINANCIAL CORPORATION,TX,75230,None,Consent not provided,Web,05/08/2015,Closed with non-monetary relief,Yes,No,1365000
655503,08/25/2016,Debt collection,"Other (i.e. phone, health club, etc.)",Cont'd attempts collect debt not owed,Debt is not mine,None,None,"Dynamic Recovery Solutions, LLC",MA,01012,"Older American, Servicemember",Consent not provided,Web,08/25/2016,Closed with non-monetary relief,Yes,No,2082447
490591,05/30/2017,Debt collection,Other debt,Attempts to collect debt not owed,Debt is not yours,"Reviewed My XXXX , XXXX , and XXXX repor...",None,ConServe,SC,294XX,None,Consent provided,Web,05/30/2017,Closed with explanation,Yes,N/A,2496972
548952,02/25/2017,Debt collection,I do not know,Taking/threatening an illegal action,Threatened arrest/jail if do not pay,A company called Fast Cash Services claims tha...,None,"COMMUNITY CHOICE FINANCIAL, INC.",WI,544XX,None,Consent provided,Web,03/08/2017,Closed with explanation,Yes,No,2361071
85062,01/02/2019,Credit card or prepaid card,Store credit card,Getting a credit card,Sent card you never applied for,"Summary : What I was told was a "" re-authoriza...",Company has responded to the consumer and the ...,"CITIBANK, N.A.",MD,208XX,None,Consent provided,Web,01/02/2019,Closed with non-monetary relief,Yes,N/A,3114520


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1282355 entries, 0 to 1282354
Data columns (total 18 columns):
 #   Column                   Non-Null Count    Dtype 
---  ------                   --------------    ----- 
 0   date_received            1282355 non-null  object
 1   product                  1282355 non-null  object
 2   sub_product              1047189 non-null  object
 3   issue                    1282355 non-null  object
 4   sub_issue                751169 non-null   object
 5   consumer_narrative       383564 non-null   object
 6   company_public_response  449082 non-null   object
 7   company                  1282355 non-null  object
 8   state                    1262955 non-null  object
 9   zip_code                 1167057 non-null  object
 10  tags                     175643 non-null   object
 11  consumer_consent         1252821 non-null  object
 12  submitted_via            1282355 non-null  object
 13  date_sent                1282355 non-null  object
 14  co

In [9]:
print(df.duplicated().sum())

0


In [10]:
df.isnull().sum()

date_received                    0
product                          0
sub_product                 235166
issue                            0
sub_issue                   531186
consumer_narrative          898791
company_public_response     833273
company                          0
state                        19400
zip_code                    115298
tags                       1106712
consumer_consent             29534
submitted_via                    0
date_sent                        0
company_response                 7
timely_response                  0
consumer_disputed                0
complaint_id                     0
dtype: int64

In [11]:
df.nunique()

date_received                 2717
product                         18
sub_product                     76
issue                          167
sub_issue                      218
consumer_narrative          366945
company_public_response         10
company                       5275
state                           63
zip_code                     22591
tags                             3
consumer_consent                 5
submitted_via                    6
date_sent                     2666
company_response                 8
timely_response                  2
consumer_disputed                3
complaint_id               1282355
dtype: int64

### 3.2 Filter to 3 Credit Bureaus

In [12]:
# Top 10 most complained-about companies
top_companies = df['company'].value_counts().head(10)
print(top_companies)

company
EQUIFAX, INC.                             115703
Experian Information Solutions Inc.       103784
TRANSUNION INTERMEDIATE HOLDINGS, INC.     96587
BANK OF AMERICA, NATIONAL ASSOCIATION      82104
WELLS FARGO & COMPANY                      70919
JPMORGAN CHASE & CO.                       60221
CITIBANK, N.A.                             49058
CAPITAL ONE FINANCIAL CORPORATION          34581
Navient Solutions, LLC.                    29296
OCWEN LOAN SERVICING LLC                   27750
Name: count, dtype: int64


In [13]:
# Total complaints for top 3 companies
top_3 = df['company'].value_counts().head(3)
bureau_total = top_3.sum()
total = df.shape[0]
percentage = round(bureau_total / total * 100, 1)

print(f"\nThe top 3 companies account for {bureau_total:,} complaints out of {total:,} total ({percentage}%)")


The top 3 companies account for 316,074 complaints out of 1,282,355 total (24.6%)


### Why focus on the 3 Credit Bureaus?

The data reveals that the 3 major credit bureaus, Equifax, 
Experian, and TransUnion, are the TOP 3 most complained-about 
companies in the entire dataset, accounting for 316,074 complaints 
(24.6% of all complaints).

Unlike banks which offer many services, credit bureaus have ONE job: 
maintaining accurate credit reports. Yet they receive more complaints 
than any bank. This makes them the most compelling subject for an 
accountability analysis.

In [14]:
df.shape

(1282355, 18)

In [15]:
df_bureaus = df[df['company'].isin([
    'EQUIFAX, INC.', 
    'Experian Information Solutions Inc.',
    'TRANSUNION INTERMEDIATE HOLDINGS, INC.'
])].copy()

print(df_bureaus.shape)

(316074, 18)


### 3.3 Drop Unusable Columns

The following columns are dropped due to high null rates that make them unusable for analysis:

- `consumer_narrative` : 68% null - no NLP analysis planned
- `company_public_response` : 48% null - too many missing values
- `consumer_consent` : 4% null - not relevant to our analysis
- `zip_code` : 26,605 nulls - redundant with `state` for geographic analysis

The following columns are kept despite having nulls:
- `tags` : 88% null but contains key vulnerability data (Older American, Servicemember)
- `sub_product` : 42% null but useful for product breakdown
- `sub_issue` : 1.7% null - minor, kept for detailed analysis
- `state` : 1% null - kept for geographic analysis
- `consumer_disputed` : Yes/No/N/A - kept for resolution quality analysis

After dropping, we retain 14 columns for analysis.

In [16]:
df_bureaus.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316074 entries, 41 to 1223910
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   date_received            316074 non-null  object
 1   product                  316074 non-null  object
 2   sub_product              182628 non-null  object
 3   issue                    316074 non-null  object
 4   sub_issue                310698 non-null  object
 5   consumer_narrative       99925 non-null   object
 6   company_public_response  164988 non-null  object
 7   company                  316074 non-null  object
 8   state                    312815 non-null  object
 9   zip_code                 289469 non-null  object
 10  tags                     36795 non-null   object
 11  consumer_consent         304224 non-null  object
 12  submitted_via            316074 non-null  object
 13  date_sent                316074 non-null  object
 14  company_response       

In [18]:
df_bureaus.isnull().sum().sort_values(ascending=False)

tags                       279279
consumer_narrative         216149
company_public_response    151086
sub_product                133446
zip_code                    26605
consumer_consent            11850
sub_issue                    5376
state                        3259
company_response                4
date_received                   0
product                         0
issue                           0
company                         0
submitted_via                   0
date_sent                       0
timely_response                 0
consumer_disputed               0
complaint_id                    0
dtype: int64

In [21]:
# Check unique values for columns with nulls
print("=== tags ===")
print(df_bureaus['tags'].value_counts())

print("\n=== zip_code ===")
print(df_bureaus['zip_code'].value_counts().head(10))

print("\n=== consumer_consent ===")
print(df_bureaus['consumer_consent'].value_counts())

print("\n=== consumer_disputed ===")
df_bureaus['consumer_disputed'].value_counts()


=== tags ===
tags
Servicemember                    21699
Older American                   13168
Older American, Servicemember     1928
Name: count, dtype: int64

=== zip_code ===
zip_code
330XX    2513
300XX    2404
770XX    2143
331XX    1988
303XX    1706
334XX    1604
606XX    1563
302XX    1389
333XX    1320
750XX    1256
Name: count, dtype: int64

=== consumer_consent ===
consumer_consent
Consent not provided    108312
Consent provided        100106
N/A                      89187
Other                     6254
Consent withdrawn          365
Name: count, dtype: int64

=== consumer_disputed ===


consumer_disputed
N/A    182056
No     112971
Yes     21047
Name: count, dtype: int64

In [22]:
df_bureaus = df_bureaus.drop(columns=[
    'consumer_narrative',
    'company_public_response',
    'consumer_consent',
    'zip_code'
])

print(df_bureaus.shape)
df_bureaus.columns.tolist()

(316074, 14)


['date_received',
 'product',
 'sub_product',
 'issue',
 'sub_issue',
 'company',
 'state',
 'tags',
 'submitted_via',
 'date_sent',
 'company_response',
 'timely_response',
 'consumer_disputed',
 'complaint_id']

In [24]:
df_bureaus.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316074 entries, 41 to 1223910
Data columns (total 14 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   date_received      316074 non-null  object
 1   product            316074 non-null  object
 2   sub_product        182628 non-null  object
 3   issue              316074 non-null  object
 4   sub_issue          310698 non-null  object
 5   company            316074 non-null  object
 6   state              312815 non-null  object
 7   tags               36795 non-null   object
 8   submitted_via      316074 non-null  object
 9   date_sent          316074 non-null  object
 10  company_response   316070 non-null  object
 11  timely_response    316074 non-null  object
 12  consumer_disputed  316074 non-null  object
 13  complaint_id       316074 non-null  int64 
dtypes: int64(1), object(13)
memory usage: 36.2+ MB


In [25]:
df_bureaus['date_received'] = pd.to_datetime(df_bureaus['date_received'])
df_bureaus['date_sent'] = pd.to_datetime(df_bureaus['date_sent'])

### 3.4 Handling Missing Values

After dropping unusable columns, the remaining null values are handled as follows:

**Columns with nulls we keep as-is:**
- `tags` : 279,279 nulls - null simply means the consumer has no special tag 
(neither "Older American" nor "Servicemember"). This is valid information.
- `sub_product` : 133,446 nulls - not critical for our core business questions. 
We use `product` instead.
- `sub_issue` : 5,376 nulls - not critical for our core business questions. 
We use `issue` instead.
- `state` : 3,259 nulls (1%) - minor. These rows will be automatically 
excluded when performing geographic analysis.

**Columns where we fill nulls:**
- `company_response` : 4 nulls - filled with "Unknown" to maintain 
data integrity without losing any rows.

**Key distinction:**
- `"N/A"` as text (e.g. in `consumer_disputed`) = a real value 
meaning "not applicable" : we leave it as-is

In [28]:
df_bureaus['company_response'] = df_bureaus['company_response'].fillna('Unknown')

In [29]:
df_bureaus.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316074 entries, 41 to 1223910
Data columns (total 14 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   date_received      316074 non-null  datetime64[ns]
 1   product            316074 non-null  object        
 2   sub_product        182628 non-null  object        
 3   issue              316074 non-null  object        
 4   sub_issue          310698 non-null  object        
 5   company            316074 non-null  object        
 6   state              312815 non-null  object        
 7   tags               36795 non-null   object        
 8   submitted_via      316074 non-null  object        
 9   date_sent          316074 non-null  datetime64[ns]
 10  company_response   316074 non-null  object        
 11  timely_response    316074 non-null  object        
 12  consumer_disputed  316074 non-null  object        
 13  complaint_id       316074 non-null  int64      

### 3.5 Normalize product names

### 3.6 Normalize issue names